File: Python/Graph/Count occurrences of a given word in a 2-d array.py
#Problem link:https://practice.geeksforgeeks.org/problems/count-occurences-of-a-given-word-in-a-2-d-array/1

#Time Complexity: O(R*C*2^(2*len)) where len is the length of the target string
#Space Compleity: O(1)
class Solution:
    
    def dfs(self,array,target,i,j,idx):
        count = 0
        #if the elements has the correct boundaries
        #and equals the character from the target string
        if(i>=0 and j>=0 and i<len(array) and j<len(array[0]) and array[i][j] == target[idx]):
            #if character found in target
            #add a star to mark it 
            array[i][j] = '*'
    
            if(idx == len(target)-1):
                count = 1
            else: #go in each possible direction
                  #North, South, East, West
                  #to look for the next character
                  #from the target string
                count += Solution().dfs(array,target,i+1,j,idx+1)
                count += Solution().dfs(array,target,i-1,j,idx+1)
                count += Solution().dfs(array,target,i,j+1,idx+1)
                count += Solution().dfs(array,target,i,j-1,idx+1)
            
            #After the recursion is done
            #put the character back in the matrix
            array[i][j] = target[idx]
        return count
    
    
    def findOccurrence(self,mat,target):
        
        occurences=0
        n = len(mat)
        m = len(mat[0])
        for i in range(n):
            for j in range(m):
                #looking for occurences starting
                #from each position
                occurences += Solution().dfs(mat,target,i,j,0)       
        
        return occurences

File: Java/Graph/Count occurrences of a given word in a 2-d array.java
// problem Link: https://practice.geeksforgeeks.org/problems/count-occurences-of-a-given-word-in-a-2-d-array/1

/****
 * Approach: we are going through one by one each element of given 2d matrix.
 * if any element is matches with target string first character,
 * it means we run DFS for making target word with this starting character
 * of 2d array. and if word is possible then we increment count
 * 
 * Total function:
 * 1. isValid: confirms that particular index is inside in 2d array.
 * 2. isDFS: use DFS Approach to make word by using characters of 2d array.
 * 3. findOccurrence: check for starting character of target string and increment
 * count if is possible to make word.
 */


/****
 * Time Complexity: O(R*C*2*(2*len)), Where len is length of target String.
 * Space Complexity: O(1)
 */


class Solution
{
    static int dx[]={1,0,-1,0};
    static int dy[]={0,1,0,-1};
    static int count=0;
    public static boolean isValid(int n,int m,int i,int j){
        if(i<0 || j<0 || i>=n || j>=m){
            return false;
        }
        return true;
    }
    public static void isDFS(char mat[][],int i,int j,int n,int m,
    String target,int idx){
        if(idx==target.length()){
            count++;
            return;
        }
        for (int p=0;p<4;p++){
            int X=dx[p]+i;
            int Y=dy[p]+j;
            if(isValid(n,m,X,Y) && mat[X][Y]==target.charAt(idx)){
                char temp=mat[i][j];
                mat[i][j]='*';
                isDFS(mat,X,Y,n,m,target,idx+1);
                mat[i][j]=temp;
            }
        }
    }
    public int findOccurrence(char mat[][], String target)
    {
        // Write your code here
        //int count=0;
        count=0;
        int n=mat.length,m=mat[0].length;
        for (int i=0;i<n;i++){
            for (int j=0;j<m;j++){
                if(mat[i][j]==target.charAt(0)){
                    //boolean visited[][]=new boolean[n][m];
                    isDFS(mat,i,j,n,m,target,1);
                }
            }
        }
        return count;
    }
}
